jib-core/src/main/java/com/google/cloud/tools/jib/http/Connection.java

```diff
@@ -114,87 +109,157 @@ private static void addProxyCredentials(ApacheHttpTransport transport, String pr
+    this.enableHttpAndInsecureFailover = enableHttpAndInsecureFailover;
+    this.sendAuthorizationOverHttp = sendAuthorizationOverHttp;
+    this.logger = logger;
+    this.secureHttpTransportFactory = secureHttpTransportFactory;
+    this.insecureHttpTransportFactory = insecureHttpTransportFactory;
   }
 
   /**
    * Sends the request with method GET.
    *
+   * @param url endpoint URL
    * @param request the request to send
    * @return the response to the sent request
    * @throws IOException if sending the request fails
    */
-  public Response get(Request request) throws IOException {
-    return send(HttpMethods.GET, request);
+  public Response get(URL url, Request request) throws IOException {
+    return call(HttpMethods.GET, url, request);
   }
 
   /**
    * Sends the request with method POST.
    *
+   * @param url endpoint URL
    * @param request the request to send
    * @return the response to the sent request
    * @throws IOException if sending the request fails
    */
-  public Response post(Request request) throws IOException {
-    return send(HttpMethods.POST, request);
+  public Response post(URL url, Request request) throws IOException {
+    return call(HttpMethods.POST, url, request);
   }
 
   /**
    * Sends the request with method PUT.
    *
+   * @param url endpoint URL
    * @param request the request to send
    * @return the response to the sent request
    * @throws IOException if sending the request fails
    */
-  public Response put(Request request) throws IOException {
-    return send(HttpMethods.PUT, request);
+  public Response put(URL url, Request request) throws IOException {
+    return call(HttpMethods.PUT, url, request);
   }
 
   /**
    * Sends the request.
    *
    * @param httpMethod the HTTP request method
+   * @param url endpoint URL
    * @param request the request to send
    * @return the response to the sent request
    * @throws IOException if building the HTTP request fails.
    */
-  public Response send(String httpMethod, Request request) throws IOException {
-    Preconditions.checkState(httpResponse == null, "Connection can send only one request");
+  public Response call(String httpMethod, URL url, Request request) throws IOException {
+    if (!isHttpsProtocol(url) && !enableHttpAndInsecureFailover) {
+      throw new SSLException("insecure HTTP connection not allowed: " + url);
+    }
+
+    try {
+      return call(httpMethod, url, request, secureHttpTransportFactory.get());
+
+    } catch (SSLException ex) {
+      if (!enableHttpAndInsecureFailover) {
+        throw ex;
+      }
+
+      try {
+        logInsecureHttpsFailover(url);
+        return call(httpMethod, url, request, insecureHttpTransportFactory.get());
+
+      } catch (SSLException ignored) { // This is usually when the server is plain-HTTP.
+        logHttpFailover(url);
+        return call(httpMethod, toHttp(url), request, secureHttpTransportFactory.get());
+      }
```

Comment:
So the problem here (and it was there before) is that we lose the SSL exception which is useful for tracking problems.

Reply:
It has made diagnosis of misconfigured registries issues difficult though.  I remember wondering why there was no PKIX exception being reported.  Maybe the right thing here is to have specific exceptions that records the cascade:
  - unable to reach via SSL: exception
  - unable to reach by plain HTTP: exception

And you know, perhaps if `enableHttpAndInsecureFailover == true`, then we should use the `insecureHttpTransportFactory` in the first place.

Reply:
The second exception can never be a PKIX exception (that can be thrown when the server and the client can at least talk on TLS). Perhaps you are misremembering something. I don't recall any particular issue from this. And this `SSLException` almost all the time means the endpoint is not TLS except for serious I/O errors, I honestly think surfacing `SSLException` will cause more confusion most of the time. And this failover will happen a lot with a local HTTP Docker registry. The path is also similar with `ConnectException` when the port is not given. The first `ConnectException` is not recorded for the port 443.

And note we are recording logs when we failover, so I think we are covered. And I am going to increase it to warning when I cache an insecure registry list, and it will become very obvious.

For these reasons, I'd like to retain the current behavior. It was actually my reasoning when I first implemented this, but then the log level was decreased to INFO.

And for now, we should always try the secure transport first (if I understood what you said correctly); some registries may be secure and others may not.

Reply:
I was thinking some more and I agree with you: with the current allowInsecureRegistries=true/false, we need to do full verification and then failover to no-validate and then plain HTTP.  When we allow insecure-registries on a per-host basis, we can skip to just first disable cert validation.

(BTW: I was referring to the original `SSLException ex` from line 223.  But unless we have allowInsecureRegistries==true then it should be thrown…. Hmm.)